File: samples/java/ShoppingMall/Shared/src/main/java/systems/zlink/samples/shoppingmall/shared/contracts/Messages.java
package systems.zlink.samples.shoppingmall.shared.contracts;

import java.math.BigDecimal;
import java.util.List;

public final class Messages {
    private Messages() {}

    public static final class OrderStatuses {
        public static final String Created = "Created";
        public static final String InventoryReserved = "InventoryReserved";
        public static final String PaymentAuthorized = "PaymentAuthorized";
        public static final String Confirmed = "Confirmed";
        public static final String Failed = "Failed";

        private OrderStatuses() {}
    }

    public record StartOrderReq(
            String cartId,
            String shippingAddressId,
            String paymentMethodId,
            String idempotencyKey) {}

    public record StartOrderRes(String orderId, String status) {}

    public record GetOrderStateReq(String orderId) {}

    public record GetOrderStateRes(OrderState state) {}

    public record OrderState(
            String orderId,
            String status,
            String shippingAddressId,
            String reservationId,
            String paymentId,
            String reason,
            BigDecimal amount,
            String currency,
            long updatedAtUnixMs) {}

    public record OrderLineInput(String sku, int quantity) {}

    public record CartSeed(
            String cartId, List<OrderLineInput> lines, BigDecimal amount, String currency) {}

    public record InventorySeed(String sku, int availableQuantity) {}

    public record PaymentMethodSeed(
            String paymentMethodId, boolean shouldAuthorize, String failureReason) {}

    public record StartOrderWorkflowReq(
            String orderId,
            String cartId,
            String shippingAddressId,
            String paymentMethodId,
            String idempotencyKey,
            List<OrderLineInput> lines,
            BigDecimal amount,
            String currency) {}

    public record StartOrderWorkflowRes(OrderState state) {}

    public record ContinueOrderWorkflowReq(String orderId) {}

    public record RunOrderWorkflowMsg(String orderId) {}

    public record ContinueOrderWorkflowRes(OrderState state, long objectGeneration) {}

    public record RebuildOrderProjectionReq(String orderId) {}

    public record RebuildOrderProjectionRes(OrderState state) {}

    public record PrepareInventoryReservedCheckpointReq(StartOrderWorkflowReq request) {}

    public record ServerAssertionReq(
            String successfulOrderId,
            String pendingRecoveredOrderId,
            String concurrentOrderId,
            String resumedOrderId,
            String inventoryFailureOrderId,
            String paymentFailureOrderId,
            String scaleOutOrderId) {}

    public record ServerAssertionRes(boolean passed, List<String> evidence) {}
}

File: samples/kotlin/ShoppingMall/Shared/src/main/kotlin/systems/zlink/samples/kotlin/shoppingmall/shared/contracts/Messages.kt
package systems.zlink.samples.kotlin.shoppingmall.shared.contracts

import systems.zlink.framework.handlers.ZLinkPacket

object OrderStatuses {
    const val Created = "Created"
    const val InventoryReserved = "InventoryReserved"
    const val PaymentAuthorized = "PaymentAuthorized"
    const val Confirmed = "Confirmed"
    const val Failed = "Failed"
}

data class OrderLineInput(val sku: String, val quantity: Int)

data class OrderState(
    val orderId: String,
    val status: String,
    val shippingAddressId: String?,
    val reservationId: String?,
    val paymentId: String?,
    val reason: String?,
    val amount: Double?,
    val currency: String?,
    val updatedAtUnixMs: Long,
)

// ----- client HTTP-style contracts -----

// --8<-- [start:doc-explicit-packet-name]
@ZLinkPacket("StartOrderReq")
data class StartOrderReq(
    val cartId: String,
    val shippingAddressId: String,
    val paymentMethodId: String,
    val idempotencyKey: String,
)

// --8<-- [end:doc-explicit-packet-name]

data class StartOrderRes(val orderId: String, val status: String)

@ZLinkPacket("GetOrderStateReq") data class GetOrderStateReq(val orderId: String)

data class GetOrderStateRes(val state: OrderState)

@ZLinkPacket("DeleteProjectionReq") data class DeleteProjectionReq(val orderId: String)

data class DeleteProjectionRes(val deleted: Boolean)

@ZLinkPacket("RebuildProjectionApiReq") data class RebuildProjectionApiReq(val orderId: String)

data class RebuildProjectionApiRes(val state: OrderState)

@ZLinkPacket("PrepareInventoryReservedApiReq")
data class PrepareInventoryReservedApiReq(val request: StartOrderReq)

data class PrepareInventoryReservedApiRes(val state: OrderState, val objectGeneration: Long)

@ZLinkPacket("PrepareInventoryReservedReq")
data class PrepareInventoryReservedReq(val command: StartOrderWorkflowReq)

data class PrepareInventoryReservedRes(val state: OrderState, val objectGeneration: Long)

@ZLinkPacket("CreatePendingMappingReq")
data class CreatePendingMappingReq(
    val idempotencyKey: String,
    val orderId: String,
    val ownerInstanceId: String,
)

data class CreatePendingMappingRes(val created: Boolean)

@ZLinkPacket("ServerAssertionReq")
data class ServerAssertionReq(
    val successfulOrderId: String,
    val pendingRecoveredOrderId: String,
    val concurrentOrderId: String,
    val resumedOrderId: String,
    val inventoryFailureOrderId: String,
    val paymentFailureOrderId: String,
    val scaleOutOrderId: String,
)

data class ServerAssertionRes(val passed: Boolean, val evidence: List<String>)

// ----- workflow command contracts (CommerceApi -> OrderWorkflow) -----

@ZLinkPacket("StartOrderWorkflowReq")
data class StartOrderWorkflowReq(
    val orderId: String,
    val cartId: String,
    val shippingAddressId: String,
    val paymentMethodId: String,
    val idempotencyKey: String,
    val lines: List<OrderLineInput>,
    val amount: Double,
    val currency: String,
)

data class StartOrderWorkflowRes(val state: OrderState)

@ZLinkPacket("ContinueOrderWorkflowReq") data class ContinueOrderWorkflowReq(val orderId: String)

data class RunOrderWorkflowMsg(val orderId: String)

data class ContinueOrderWorkflowRes(val state: OrderState, val objectGeneration: Long? = null)

@ZLinkPacket("RebuildOrderProjectionReq") data class RebuildOrderProjectionReq(val orderId: String)

data class RebuildOrderProjectionRes(val state: OrderState)

// ----- self-check seed contracts -----

data class CartSeed(
    val cartId: String,
    val lines: List<OrderLineInput>,
    val amount: Double,
    val currency: String,
)

data class InventorySeed(val sku: String, val availableQuantity: Int)

data class PaymentMethodSeed(
    val paymentMethodId: String,
    val shouldAuthorize: Boolean,
    val failureReason: String?,
)
